### PR TITLE
Transformer End-to-End Speech Translation

### PR DESCRIPTION
**Pull request by hirofumi0810**
- Add ST task-specific Transformer
- Support ASR multitask (CTC, attention, CTC+Attention)
- Support MT multitask
- Update RNN model following Transformer
- Remove model dependencies from Transformer ASR
- Enhance test for Transformer ASR

### CODE REVIEW

**Comment by sw005320** [comment]
@butsugiri, could you review this PR?

File: espnet/nets/pytorch_backend/e2e_st.py [hunk 1]
```diff
@@ -266,56 +291,87 @@ def forward(self, xs_pad, ilens, ys_pad, ys_pad_asr):
...
+            cers = []
+
+            y_hats = self.ctc.argmax(hs_pad).data
+            for i, y in enumerate(y_hats):
+                y_hat = [x[0] for x in groupby(y)]
+                y_true = ys_pad_asr[i]
 
-        # 4. compute cer/wer
-        if self.asr_weight == 0:
+                seq_hat = [self.char_list[int(idx)] for idx in y_hat if int(idx) != -1]
```
Comment (Fhrozen) [inline espnet/nets/pytorch_backend/e2e_st.py]: Probably, you could try with https://github.com/espnet/espnet/blob/f10be2416bf14bdc33d00313b95d2944f954ecc5/espnet/nets/e2e_asr_common.py#L88

Comment (hirofumi0810) [reply espnet/nets/pytorch_backend/e2e_st.py]: @Fhrozen You are right. But the RNN model has a beam search option during training while Transformer allows greedy decoding only. That's why these interfaces for the WER calculation are different. Maybe it's better to integrate them in another PR.

File: espnet/bin/st_train.py [hunk 2]
```diff
@@ -71,11 +74,15 @@ def get_parser(parser=None, required=True):
     parser.add_argument('--model-module', type=str, default=None,
                         help='model defined module (default: espnet.nets.xxx_backend.e2e_st:E2E)')
     # loss related
+    parser.add_argument('--ctc_type', default='warpctc', type=str,
+                        choices=['builtin', 'warpctc'],
+                        help='Type of CTC implementation to calculate loss.')
     parser.add_argument('--mtlalpha', default=0.0, type=float,
-                        help='Multitask learning coefficient, alpha: alpha*ctc_loss + (1-alpha)*att_loss ')
-    # TODO(hirofumi0810): remove this after adding e2e_st_transformer.py
+                        help='Multitask learning coefficient, alpha: \
```
Comment (sw005320) [inline espnet/bin/st_train.py]: did you decide to keep it?

File: espnet/nets/pytorch_backend/e2e_st_transformer.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
+# Copyright 2019 Shigeki Karita
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_st_transformer.py]: Maybe you can add your name.
Also, could you add some comments which part mainly you modified/added from the original ASR transformer code (of course translation method, but are there anything else)?

File: espnet/bin/st_train.py [hunk 4]
```diff
@@ -300,4 +306,10 @@ def main(cmd_args):
 
 
 if __name__ == '__main__':
+    # NOTE(kan-bayashi): setting multiple times causes RuntimeError
+    #   See also https://github.com/pytorch/pytorch/issues/3492
+    try:
+        mp.set_start_method('spawn')
+    except RuntimeError:
+        pass
```
Comment (kan-bayashi) [inline espnet/bin/st_train.py]: Due to #1290, I decided to remove these lines.

File: espnet/bin/st_train.py (hunk 2 above)
Comment (hirofumi0810) [reply espnet/bin/st_train.py]: Yes, this is used for weighting CTC loss.

File: espnet/asr/pytorch_backend/asr.py [hunk 5]
```diff
@@ -167,8 +168,10 @@ def update_core(self):
 
         # Get the next batch ( a list of json files)
         batch = train_iter.next()
-        # self.iteration += 1 # Increase may result in early report, which is done in other place automatically.
-        x = tuple(arr.to(self.device) for arr in batch)
+        self.iteration += 1
```
Comment (sw005320) [inline espnet/asr/pytorch_backend/asr.py]: could you tell me why you revert `self.iteration += 1`?

File: espnet/asr/pytorch_backend/asr.py [hunk 6]
```diff
@@ -167,8 +168,10 @@ def update_core(self):
...
+
+        x = tuple(arr.to(self.device) if arr is not None else None
```
Comment (sw005320) [inline espnet/asr/pytorch_backend/asr.py]: Can you leave the comment for this change?
What is the intention?

**Comment by sw005320** [comment]
This is an example script
https://github.com/espnet/espnet/blob/master/egs/tedlium2/asr1/local/recog_wav.sh
I'm not sure how it is maintained. 
@takenori-y, could you follow up it?

The actual implementation is around https://github.com/espnet/espnet/blob/2d2151de1262d2fca375d71bbcc4ef1b8570723c/espnet/nets/pytorch_backend/streaming/segment.py#L40

File: espnet/asr/pytorch_backend/asr.py (hunk 5 above)
Comment (hirofumi0810) [reply espnet/asr/pytorch_backend/asr.py]: This is my mistake.

File: espnet/asr/pytorch_backend/asr.py (hunk 6 above)
Comment (hirofumi0810) [reply espnet/asr/pytorch_backend/asr.py]: ditto

File: espnet/asr/pytorch_backend/asr.py (hunk 5 above)
Comment (hirofumi0810) [reply espnet/asr/pytorch_backend/asr.py]: Maybe I uncommented it during fixing some conflicts by mistake.

File: espnet/nets/pytorch_backend/e2e_st_transformer.py (hunk 3 above)
Comment (hirofumi0810) [reply espnet/nets/pytorch_backend/e2e_st_transformer.py]: ASR/MT multi-task only.

**Comment by hirofumi0810** [comment]
All done. Ready to be merged.

File: espnet/nets/pytorch_backend/e2e_st.py (hunk 1 above)
Comment (sw005320) [reply espnet/nets/pytorch_backend/e2e_st.py]: I see.

**Comment by sw005320** [comment]
This is a great function!
Thanks a lot!

**Pull request merged**